### PR TITLE
Variable weights on Line Items

### PR DESCRIPTION
**Pull request by lin-d-hop**
Hi guys!

Assigning to Rob as I remember Kirsten saying that Rob can check this one over...

I do have one or two questions on the specs:

1) spec/controllers/spree/api/line_items_controller_spec.rb
The spec really only checks that the after filter is called. I found this link http://axonflux.com/rspecs-shouldreceive-doesnt-ac saying that should_receive only mocks and therefore can't itself be used to test order.update_distribution_charge! is received... which is probably more the behaviour I want.
Should I now be writing a test to check that apply_enterprise_fees actually calls order.update_distribution_charge! ? As it is basically a one line method that seems a little OTT.

2) Do I need to add specs testing that order.update_distribution_charge! performs the same function with when the cart is completed. Looking at the existing functionality and specs the state of the cart makes no difference to functionality so it feels like I would simply be repeating the existing specs, which seems a bit wasteful.

Basically in thinking about this I was trading off between philosophies of 'specs for the sake of specs' and being super thorough. So some input into these questions would be appreciated.

Thanks for your input. Sorry it's taken me a while to get this up to this stage. I've had soooo much to catch up on after my time away... easy to underestimate these things :-)

Hope all is good in the hood xxx

### CODE REVIEW

**Comment by lin-d-hop** [comment]
@oeoeaio just wondering if you saw this pull request and have had a moment to review at all.

@kirstenalarsen is this still ok to assign to Rob? In your expert project management hands :-)

**Comment by oeoeaio** [comment]
@lin-d-hop Yep! Sorry! Had a bit of a full-on week. I'll review asap. Maybe on the train tomorrow morning.

**Comment by lin-d-hop** [comment]
@oeoeaio Thanks! I know you guys are busy beavers so sorry to bump. I really appreciate your help/efforts dude :-)

File: app/views/spree/api/line_items/bulk_show.v1.rabl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 object @line_item
-attributes :id, :quantity, :max_quantity
+attributes :id, :quantity, :max_quantity, :price
```
Comment (oeoeaio) [inline app/views/spree/api/line_items/bulk_show.v1.rabl]: I don't think we need :price here when we have a node named :price below? I could be wrong

File: spec/controllers/spree/api/line_items_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    #test that when a line item is updated, an order's fees are updated too
+    context "as an admin user" do
+      sign_in_as_admin!
+      make_simple_data!
+
+      context "as a line item is updated" do
+        it "apply enterprise fees on the item" do
+          line_item_params = { order_id: order.number, id: line_item.id, line_item: { id: line_item.id, unit_value: 520 }, format: :json}
+          controller.should_receive(:apply_enterprise_fees).and_return(true)
```
Comment (oeoeaio) [inline spec/controllers/spree/api/line_items_controller_spec.rb]: I think here you can probably instead test that your order receives update_distribution_charge!. Testing that the after_filter is called is testing the rails is testing that the rails source code is working, rather than our own. What we actually want to know is that every time run the update action on the line item api controller, update_distribution_charge! is called on the corresponding order, so that is the behaviour we should test.

**Comment by oeoeaio** [comment]
With regard to your second question, I don't think you need to test the behaviour of the update_distribution_charge! method, no. We have specs which test the expected behaviour of this method, which should be written in a context independent way, so unless we change the behaviour of the method itself, I don't think we need to write a new spec just because it is used in a new context.

File: app/views/spree/admin/orders/bulk_management.html.haml [hunk 3]
```diff
@@ -100,53 +100,60 @@
...
+            %td.variant{ 'ng-show' => 'columns.variant.visible' }
+              %a{ :href => '#', 'ng-click' => "setSelectedUnitsVariant(line_item.units_product,line_item.units_variant)" } {{ line_item.units_variant.unit_text }}
+            %td.quantity{ 'ng-show' => 'columns.quantity.visible' }
+              %input{ :type => 'number', :name => 'quantity', 'ng-model' => "line_item.quantity", 'ofn-line-item-upd-attr' => "quantity" }
+            %td.max{ 'ng-show' => 'columns.max.visible' } {{ line_item.max_quantity }}
+            %td.unit_value{ 'ng-show' => 'columns.unit_value.visible' }
+              %input{ :type => 'number', :name => 'unit_value', :id => 'unit_value', 'ng-model' => "line_item.unit_value", 'ng-change' => "weightAdjustedPrice(line_item, {{ line_item.unit_value }})", 'ofn-line-item-upd-attr' => "unit_value" }
+            %td.price{ 'ng-show' => 'columns.price.visible' }
+              %input{ :type => 'text', :name => 'price', :id => 'price', :value => '{{ line_item.price | currency }}', 'ng-model' => "line_item.price", 'ng-readonly' => "true", 'ofn-line-item-upd-attr' => "price" }
```
Comment (oeoeaio) [inline app/views/spree/admin/orders/bulk_management.html.haml]: This needs a bit of work, I think something weird is happening with having the value set in two places: ng-model is trying to set the value to one thing, and then the dynamic setting of the value with {{ line_item.price | currency }} is trying to set it to something else. I am going to have a look at tweaking the directive so that it doesn't depend on ng-model to work, which will mean that we can pull that out and even switch the price to a span element rather than a text field, which might be better for usability.

File: spec/controllers/spree/api/line_items_controller_spec.rb (hunk 2 above)
Comment (lin-d-hop) [inline spec/controllers/spree/api/line_items_controller_spec.rb]: Thanks @oeoeaio. I think that actually I can't test this though. I was stuck on it for ages until I found this blog post:
http://axonflux.com/rspecs-shouldreceive-doesnt-ac
Which suggests that all method calls within the stated method call will be mocked. Therefore order.update_distribution_charge! will never actually be called because all method calls within update are mocked. The post suggests you can only test should_receive on step down the stack. Hence the current test. If you have different experience then please correct me!

Comment (oeoeaio) [inline spec/controllers/spree/api/line_items_controller_spec.rb]: Hi @lin-d-hop,

Yep, basically as soon as you stub a method with should_receive, it gets replaced with a spy which listens for calls, and no code within the method is run. You can allow a pass through by adding .and_call_original to the stub. Of course you can't add a return value when you use .and_call_original.

What I was trying to say in the original message is that I don't think we want to stub apply_enterprise_fees at all, because testing that it is just testing that rails is doing what rails should do. ie. calling a method specified in an after filter at the appropriate time. Testing that is not our responsibility. The behaviour we are actually interested in testing is that .update_distribution_charge! is called on the order after the update action.

So, I think you should remove the stub (should_receive) on the apply_enterprise_fees method on the controller, and instead stub the .update_distribution_charge! method on the order, and check that that method is called when you call the line_item update action.

Something like this:
order.should_receive(:update_distribution_charge!)

Or, if you wanted to use the newer RSpec syntax:
expect(order).to receive(:update_distribution_charge!)

File: app/views/spree/admin/orders/bulk_management.html.haml (hunk 3 above)
Comment (oeoeaio) [inline app/views/spree/admin/orders/bulk_management.html.haml]: Ok, ignore me. My ofn-line-item-upd-attr directive does need a bit of work, but it shouldn't be affecting the functionality of this page, I'll keep working on it and send you something seen.

**Comment by oeoeaio** [comment]
I think I am having some issues related to all of the line items in the page having a unit_value of 0 when they are loaded into the page. I think we really need a migration which adds a unit value to all of the line items already in the database, and also some code to add the unit_value to newly created line_items, based on their variant. Do you have any strong feelings about that? That is the ultimate intention right? That all line_items should store the unit_value of the variant that they were originally associated with, but that this value can be altered at a later time using BOM?

File: app/assets/javascripts/admin/bulk_order_management.js.coffee [hunk 4]
```diff
@@ -162,6 +163,12 @@ angular.module("ofn.admin").controller "AdminOrderMgmtCtrl", [
       $scope.supplierFilter = $scope.suppliers[0].id
       $scope.orderCycleFilter = $scope.orderCycles[0].id
       $scope.quickSearch = ""
+
+    $scope.weightAdjustedPrice = (lineItem, oldValue) ->
+      if lineItem.unit_value <= 0
+        lineItem.unit_value = lineItem.units_variant.unit_value
+      lineItem.price = lineItem.price * lineItem.unit_value / oldValue
```
Comment (oeoeaio) [inline app/assets/javascripts/admin/bulk_order_management.js.coffee]: I think this really needs to be able to handle the zero unit_values. At the moment it just seems to crash every time I try to change the unit_value of a line_item from 0 to 1, because it is trying to determine the new price by dividing lineItem.unit_value by oldValue, which is zero.

Comment (lin-d-hop) [inline app/assets/javascripts/admin/bulk_order_management.js.coffee]: Oh yeah, good point. I had this logic and then I must have 'cleaned it up'...
Adding back in now. I think this is a better solution than populating on a migration... do you think?

File: app/assets/javascripts/admin/bulk_order_management.js.coffee [hunk 5]
```diff
@@ -162,6 +163,14 @@ angular.module("ofn.admin").controller "AdminOrderMgmtCtrl", [
...
       $scope.quickSearch = ""
+
+    $scope.weightAdjustedPrice = (lineItem, oldValue) ->
+      if oldValue <= 0
+        oldValue = lineItem.units_variant.unit_value
+      if lineItem.unit_value <= 0
+        lineItem.unit_value = lineItem.units_variant.unit_value
+      lineItem.price = lineItem.price * lineItem.unit_value / oldValue
+      #$scope.bulk_order_form.line_item.price.$setViewValue($scope.bulk_order_form.line_item.price.$viewValue)
 ]
```
Comment (oeoeaio) [inline app/assets/javascripts/admin/bulk_order_management.js.coffee]: Ok, don't we need to make sure that oldValue is not zero? I think that is what is causing the issue in my system.

With respect to migrating the data, I suppose we need to have a discussion about how exactly the unit_value field on line_item will be used. If we are going to need it in other pages (say reports), and we are NOT going to require that every line item has a unit_value, then we are going to have to write some logic to work out which unit_value is reported (that of the line_item itself, or that of its variant).

I think that based on my (probably limited) understanding of what this field is doing, my strong preference would be for us to have confidence that we can always read the unit_value from the line_item, and that will give us all of the information we need to know without ever having to look at the original variant. That way our data would behave consistently and in a predictable manner. I also like it because it more of less copies the pattern of the price field on line_item, which is completely abstracted from the price of the original variant and is used as the source of truth about the price the line_item was sold for.

I am happy to write the migration - I think it will just be a matter of cycling through all of the line items in the system and copying the unit value across. If you have a strong feeling another way, let me know. :)

Comment (lin-d-hop) [inline app/assets/javascripts/admin/bulk_order_management.js.coffee]: We dowant to make sure old value isn't zero.
If it is it means the weight has never been adjusted on the line item. So if it is we can safely take the default from the product variant.
A migration would do the same thing as this code, but updating all at once. This way we only populate the field if the unit_value is different from the product variant. I feel this is a valid solution as a line-items unit_value will only occasionally be modified from the variant. Does that sound reasonable?

**Comment by kirstenalarsen** [comment]
hi guys - quick question about how this will work (and something I'm hoping) . . if I come in and change the line item to zero, or delete it, will it recalculate / remove the associated adjustments? please say yes?

File: app/assets/javascripts/admin/bulk_order_management.js.coffee (hunk 5 above)
Comment (oeoeaio) [inline app/assets/javascripts/admin/bulk_order_management.js.coffee]: Hmmmmm, it makes me uncomfortable. What I am advocating a migration for existing AND some logic that sets the unit_value of every LineItem as it is created, so that we maintain data integrity. Sorry to be a pain, my issue is that the logic of "if the unit_value of the line_item is zero then use the unit_value of the variant, otherwise unit_value of the line_item" seems overly complicated and means that we will require knowledge of this particular pattern in every instance that we want to use the unit_value of a line_item. I can't really see the logic in storing a unit_value of zero, when that is not the unit value. If we are going to go to the trouble of having a field for unit_value on line_item, I think it may as well just be populated with the actual unit_value, and then our interactions with it can be intuitive, consistent and error free.

**Comment by lin-d-hop** [comment]
@kirstenalarsen Yes, that's how it works.

**Comment by kirstenalarsen** [comment]
yay :)

**Comment by lin-d-hop** [comment]
@oeoeaio To be honest we'll have to keep the logic anyway, as it does need to be able to handle the case of the unit_value being deleted by the user and then the price being set to zero accidentally.

I'm happy to add the functionality to the creation of a line_item and the migration so that the integrity is maintained.

File: spec/controllers/spree/api/line_items_controller_spec.rb (hunk 2 above)
Comment (lin-d-hop) [inline spec/controllers/spree/api/line_items_controller_spec.rb]: HI @oeoeaio 
I'm a bit stuck on this one. I added .and_call_original to the stub. And I changed to order.should_receive(:update_distribution_charge!)

When I binding.pry spree_post :update, line_item_params the correct page is def returned but the test doesn't pass. order.update_distribution_charge! is never called. 

The tinterwebs seems to suggest that and_call_original is only available on partial doubles, suggesting to me that actually this isn't genuinely calling the original method and whole stack. 

So yeah, stuck. Do you have an example of using should_receive to test a method call within a stack? All the egs I've found in the code are same layer.

Hope you're enjoying your travels. Pity you didn't make it to Farm hack. Serverine from Greenhorns was there... and it was an wicked gathering :-) Any UK time planned?

Comment (oeoeaio) [inline spec/controllers/spree/api/line_items_controller_spec.rb]: Hi @lin-d-hop,

I really think that you can just ditch the apply_enterprise_fees stub on the controller, and just have one for update_distribution_charge! on the order.

If that doesn't work, maybe push what you have into this branch and I'll have a play around and see if I can understand what is happening.

**Comment by lin-d-hop** [comment]
Hi @oeoeaio. Thanks for offering to take a look at my broken spec. I'm stuck! Committed in d344c3d

Also committed the changes for data integrity of the line_item in 6045283. Let me know if you have suggestions on that as well.

Thanks so much for you help!

**Comment by lin-d-hop** [comment]
Hi @oeoeaio 

So commit 3179887 adds the changes to not allow editing of line_item.unit_value if variant.unit_value is zero.

Let me know if anything else needed on this.

Cheers,
Lynne

**Comment by lin-d-hop** [comment]
Hi @kirstenalarsen, Just thought about your question more:  "if I come in and change the line item to zero, or delete it, will it recalculate / remove the associated adjustments? please say yes?"

If you update to 0 it'll work. If you delete it won't recalculate the fees.
It might not be too hard to make it recalculate fees on delete... is that worth looking into at some point? How urgent?
